File: 1367.linked-list-in-binary-tree.cpp
#
# @lc app=leetcode id=1367 lang=cpp
#
# [1367] Linked List in Binary Tree
#
# @lc code=start
/**
* Definition for singly-linked list.
* struct ListNode {
*     int val;
*     ListNode *next;
*     ListNode() : val(0), next(nullptr) {}
*     ListNode(int x) : val(x), next(nullptr) {}
*     ListNode(int x, ListNode *next) : val(x), next(next) {}
* };
*/
/**
* Definition for a binary tree node.
* struct TreeNode {
*     int val;
*     TreeNode *left;
*     TreeNode *right;
*     TreeNode() : val(0), left(nullptr), right(nullptr) {}
*     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
*     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
* };
*/
class Solution {
public:
    bool isSubPath(ListNode* head, TreeNode* root) {
        if (!root) return false;
        // Try to match from current node, or from left/right subtree
        return dfs(head, root) || isSubPath(head, root->left) || isSubPath(head, root->right);
    }
    
private:
    bool dfs(ListNode* head, TreeNode* root) {
        // If we've matched all of the linked list, success
        if (!head) return true;
        // If tree node is null but list isn't finished, no match
        if (!root) return false;
        // Values must match, and rest of list must match in left or right subtree
        if (head->val != root->val) return false;
        return dfs(head->next, root->left) || dfs(head->next, root->right);
    }
};
# @lc code=end

File: 1367.linked-list-in-binary-tree.py
#
# @lc app=leetcode id=1367 lang=python3
#
# [1367] Linked List in Binary Tree
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSubPath(self, head: Optional[ListNode], root: Optional[TreeNode]) -> bool:
        def helper(tnode: Optional[TreeNode], lnode: Optional[ListNode]) -> bool:
            if not lnode:
                return True
            if not tnode or tnode.val != lnode.val:
                return False
            return helper(tnode.left, lnode.next) or helper(tnode.right, lnode.next)

        def traverse(tnode: Optional[TreeNode]) -> bool:
            if not tnode:
                return False
            if helper(tnode, head):
                return True
            return traverse(tnode.left) or traverse(tnode.right)

        return traverse(root)

# @lc code=end